tests/unit/executors/indexers/test_numpyindexer.py

```diff
@@ -236,3 +236,126 @@ def test_indexer_zeros(metric, dimension, test_metas):
+    random_data = np.random.random([1, num_dim])
+
+    with BaseIndexer.load(save_abspath) as indexer:
+        assert isinstance(indexer, NumpyIndexer)
+        # this will fail cause the key doesn't exist
+        with pytest.raises(KeyError):
+            indexer.update(random_keys, random_data)
+        indexer.update([key_to_update], data_to_update)
+        indexer.save()
+
+    with BaseIndexer.load(save_abspath) as indexer:
+        assert isinstance(indexer, NumpyIndexer)
+        query_results = indexer.query_by_id([key_to_update])
+        assert np.array_equal(data_to_update, query_results)
+
+    # delete
+    keys_to_delete = 1
+    vec_idx_to_delete = vec_idx[:keys_to_delete]
+
+    with BaseIndexer.load(save_abspath) as indexer:
+        assert isinstance(indexer, NumpyIndexer)
+        indexer.delete(vec_idx_to_delete)
+        indexer.save()
+        assert indexer.size == len(vec_idx) - keys_to_delete
+
+    assert indexer.size == len(vec_idx) - keys_to_delete
+
+    with BaseIndexer.load(save_abspath) as indexer:
+        assert isinstance(indexer, NumpyIndexer)
+        assert indexer.size == len(vec_idx) - keys_to_delete
+        # this will fail. key doesn't exist
+        with pytest.raises(KeyError):
+            _ = indexer.query_by_id(vec_idx)
+        query_results = indexer.query_by_id(vec_idx[keys_to_delete:])
+        expected = vec[keys_to_delete:]
+        np.testing.assert_allclose(query_results, expected, equal_nan=True)
+
+
+@pytest.mark.parametrize('batch_size, compress_level', [(None, 0), (None, 1), (16, 0), (16, 1)])
+def test_numpy_indexer_known_and_delete(batch_size, compress_level, test_metas):
+    vectors = np.array([[1, 1, 1],
+                        [10, 10, 10],
+                        [100, 100, 100]])
+    keys = np.array([4, 5, 6])
+    with NumpyIndexer(metric='euclidean', index_filename='np.test.gz', compress_level=compress_level,
+                      metas=test_metas) as indexer:
+        indexer.batch_size = batch_size
+        indexer.add(keys, vectors)
+        indexer.save()
+        assert os.path.exists(indexer.index_abspath)
+        save_abspath = indexer.save_abspath
+
+    top_k = 3
+    queries = np.array([[1, 1, 1],
+                        [10, 10, 10]])
+    with BaseIndexer.load(save_abspath) as indexer:
+        assert isinstance(indexer, NumpyIndexer)
+        idx, dist = indexer.query(queries, top_k=top_k)
+        np.testing.assert_equal(idx, np.array([[4, 5, 6], [5, 4, 6]]))
+        assert idx.shape == dist.shape
+        assert idx.shape == (len(queries), top_k)
+        np.testing.assert_equal(indexer.query_by_id([5, 4, 6]), vectors[[1, 0, 2]])
+
+    # update and query again
+    key_to_update = np.array([4])
+    data_to_update = np.array([[1000, 1000, 1000]])
+
+    with BaseIndexer.load(save_abspath) as indexer:
+        assert isinstance(indexer, NumpyIndexer)
+        indexer.update(key_to_update, data_to_update)
+        indexer.save()
```

Comment:
why do we call save? It will be handled in the context manager

Reply:
yes, it is here because I just copied pasted from old tests.

Reply:
will handle in different PR